### PR TITLE
Proposal for WFLY-13354 Add Galleon layers for the EJB subsystem

### PR DESCRIPTION
**Pull request by chengfang**
JIRA: https://issues.redhat.com/browse/WFLY-13354

### CODE REVIEW

**Comment by chengfang** [comment]
@yersan please review.

**Comment by darranl** [comment]
For all new layers I think we need to make sure security is considered in the requirements,

We don't want to accidentally leave anything unsecured by default but ideally if layers can avoid a dependency on any security layers that would be preferred.

Failing that for the new layers if security dependencies are required we should be preferring elytron with an option to select legacy security.  At some point in the future all legacy security related layers will be removed.

Can we please add a check for these questions in the requirements?

File: ejb/WFLY-13354-galleon-for-ejb.adoc [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+2. ejb-dist-cache. This provides the infinispan subsystem resources related to distributed ejb caching
+(i.e. the ones in standalone0ha.xml and standalone-full-ha.xml.)
+
+This pattern of "exclude an optional local caching layer and add a distributed variant" is how web session
+and jpa caching are handled in the layers for those features.
+
+TBH I don't see big FS footprint improvements coming out of these permutations. Big EJB dependencies include (in no particular order.)
+
+* IIOP. But the TM uses IIOP libs so they will be there regardless.
+* Remoting. It would be nice to be able to eliminate the need for remoting, e.g. for ejb3-lite if the user app isn't a remote ejb client. But that doesn't seem trivial and if a server is manageable remoting will be there anyway to support the CLI.
```
Comment (darranl) [inline ejb/WFLY-13354-galleon-for-ejb.adoc:88]: FY the remoting layer is going to be needing some work, I am seeing it is defaulting to a dependency on legacy security realms without setting up the correct dependencies but also needs cleaning up to use elytron.

File: ejb/WFLY-13354-galleon-for-ejb.adoc [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+2. Eliminate unnecessary exposure.
+
+3. Eliminate jars, both to save filesystem footprint and to reduce any theoretical attack surface.
+
+4. Reduce configuration clutter and unnecessary services.
+
+Based on the above factors, the following layers are to be supported:
+
+1. ejb3-lite. This is like what we provide in standalone.xml, but without the MDB instance pool or the remote connector resource.
```
Comment (darranl) [inline ejb/WFLY-13354-galleon-for-ejb.adoc:61]: Do we need the 3 in all the layer names?  This is the major version number of the specification but we don't currently plan to support any other ejb version in parallel do we?

The reason I ask is it came up late in the development of JBoss AS7 that really the subsystem should not have contained the 3 but it was too late at that point.

**Comment by darranl** [comment]
I have just submitted the following ready for review to WildFly Core to adjust the current remoting layer definition:
  https://github.com/wildfly/wildfly-core/pull/4291

File: ejb/WFLY-13354-galleon-for-ejb.adoc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+(A webservices layer would depend on this.)
+
+2. ejb3. Builds (i.e. depends) on ejb3-lite and adds the discovery subsystem, the MDB pool and the remote connector resource.
+This looks what we provide in standalone-full.xml, except no IIOP integration.
+
+3. ejb3-iiop. Builds (i.e. depends) on *ejb3-lite* and adds the iiop subsystem plus the integration resource in the EJB3 subsystem.
```
Comment (darranl) [inline ejb/WFLY-13354-galleon-for-ejb.adoc:67]: I had to pause the work but the IIOP side may need this issue resolved as the legacy security subsystem will not be available as a layer https://issues.redhat.com/browse/WFLY-13679

File: ejb/WFLY-13354-galleon-for-ejb.adoc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+This *could* depend on *ejb3* instead, but that means EJB is unnecessarily exposed over the HTTP interface and a lot of
+dependencies are brought in for messaging integration.
+
+A user could provision ejb3 + ejb3-iiop and get the equivalent to what's in standalone-full.xml.
```
Comment (darranl) [inline ejb/WFLY-13354-galleon-for-ejb.adoc:71]: We won't be supporting the legacy security subsystem in bootable jar so this equivalence may not be possible.

Instead some consideration to the default security domain may be needed.

File: ejb/WFLY-13354-galleon-for-ejb.adoc (hunk 3 above)
Comment (tommaso-borgato) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:67]: @darranl shall we mention WFLY-13679 as a blocker for the current RFE?

File: ejb/WFLY-13354-galleon-for-ejb.adoc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+. Test the Galleon layer provision.
+https://github.com/wildfly/wildfly/blob/master/testsuite/layers/src/test/java/org/jboss/as/test/layers/LayersTestCase.java[LayersTestCase]
+will be expanded to test the provision of the added layers when it is used isolated and with other layers.
+For each sort of provision, this test does the following:
+.. Verifies the provisioned modules are the expected ones.
+.. Verifies the provisioned server starts successfully.
+. Execution of existing tests on a minimum set of layers. Reuse the existing tests available on the WildFly test suite,
+which are directly testing features added by these layers, and execute them on a server installation provisioned with this layer.
+
```
Comment (tommaso-borgato) [inline ejb/WFLY-13354-galleon-for-ejb.adoc:106]: @chengfang can we have some more details here? for example security and clustering test are topics of interest

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

**Comment by chengfang** [comment]
Will take a closer look at the comments and update this analysis and test plan soon.

File: ejb/WFLY-13354-galleon-for-ejb.adoc (hunk 3 above)
Comment (darranl) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:67]: For this piece I will try and look again when I am back, but we do need to make sure any supported layers don't inadvertently pull in the legacy security subsystem - I would say if they do we should raise that as a Blocker until we can triage.

Having said that some layers will still pull in the PicketBox modules but we can slowly work to clean those up.

File: ejb/WFLY-13354-galleon-for-ejb.adoc (hunk 2 above)
Comment (chengfang) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:61]: Good point. On the other hand, I think ejb3 distinghish itself from old-style, heavy-weight ejb 2.x, so it conveys some positive connotation. I'm neutral on this naming, but if there is strong perference to drop the 3, I'm fine with renaming it.

File: ejb/WFLY-13354-galleon-for-ejb.adoc (hunk 3 above)
Comment (chengfang) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:67]: @tadamski can you take a look at the iiop-openjdk's dependency on picketbox (https://issues.redhat.com/browse/WFLY-13679)? Is its dependency on legacy security really required?

**Comment by chengfang** [comment]
> For all new layers I think we need to make sure security is considered in the requirements,
> 
> We don't want to accidentally leave anything unsecured by default but ideally if layers can avoid a dependency on any security layers that would be preferred.
> 
> Failing that for the new layers if security dependencies are required we should be preferring elytron with an option to select legacy security. At some point in the future all legacy security related layers will be removed.
> 
> Can we please add a check for these questions in the requirements?

incorporated the above comment into the analysis doc.

File: ejb/WFLY-13354-galleon-for-ejb.adoc [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+== Community Documentation
+
+Community documentation plan is adding the layer to https://docs.wildfly.org/19/Admin_Guide.html#wildfly-galleon-layers[WildFly Galleon layers]
```
Comment (tommaso-borgato) [inline ejb/WFLY-13354-galleon-for-ejb.adoc:114]: @chengfang the link should be to https://docs.wildfly.org/20/Admin_Guide.html#wildfly-galleon-layers

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: ejb/WFLY-13354-galleon-for-ejb.adoc (hunk 3 above)
Comment (tommaso-borgato) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:68]: @chengfang, @darranl, @tadamski does this conversation prevents the AD from being approved? do you think we can approve the AD and move this conversation somewhere else?

Comment (chengfang) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:68]: Let's investigate it ([WFLY-13679](https://issues.redhat.com/browse/WFLY-13679) Make legacy security optional for "org.wildfly.iiop-openjdk") separately. I think it should not be blocking this proposal.

Comment (chengfang) [reply ejb/WFLY-13354-galleon-for-ejb.adoc:68]: We'll drop ejb-iiop and iiop-openjdk layers from this feature. Once WFLY-13679 is resolved, we can then add them back.

**Pull request merged**